{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data = pd.read_excel('expression_data.xlsx', sheet_name='Data')\n",
    "try:\n",
    "    data = data.drop(columns=['StartV2', 'EndV2', 'Keeptot', 'Keeptrim', 'Id260210', 'OrigId'])\n",
    "except:\n",
    "    pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genes = (data[data['classif'].map(lambda x : x == '-')]).reset_index()\n",
    "fwd = genes[genes['Strand'].map(lambda x : x == 1)]\n",
    "rvs = genes[genes['Strand'].map(lambda x : x == -1)]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Read in B. subtilis sp. 168 genome as a string"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genome_file = open(\"bsubt168.fna\",\"r\") \n",
    "genome = genome_file.read()\n",
    "genome = genome.replace('\\n','')[71:] #remove \\n and the first part of the FASTA file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Get the reverse of a sequence of the form '[ATCG]*'\n",
    "def reverse_complement (seq):\n",
    "    conv_dic = {'A':'T', 'T':'A', 'G':'C', 'C':'G'}\n",
    "    rc = ''\n",
    "    for c in seq:\n",
    "        rc = conv_dic[c] + rc\n",
    "    return rc\n",
    "\n",
    "#Get the sequence for a given gene\n",
    "#Useful for checking if the indices in the dataframe correspond to the correct\n",
    "#Genome sequence\n",
    "def get_gene_sequence (gene_name, df, genome):\n",
    "    gene_info = df[df['Name']==gene_name]\n",
    "    start, end = (int(gene_info['StartV3']), int(gene_info['EndV3'].values))\n",
    "    if gene_info['Strand'].values == -1:\n",
    "        return reverse_complement(genome[start-1:end])\n",
    "    else:\n",
    "        return genome[start-1:end]\n",
    "\n",
    "#Return means across different trials of the same condition and return one array of expression\n",
    "#Values from all of the cells\n",
    "def merge_exp_columns (df, unique_exp_columns):\n",
    "    \n",
    "    averaged_exps_df = pd.DataFrame()\n",
    "    for exp in unique_exp_columns:\n",
    "        averaged_exps_df[exp] = df.filter(regex=(exp.replace('+', '\\+').replace('.','\\.'))).mean(axis=1)\n",
    "    return pd.DataFrame({'exps' : list(averaged_exps_df.to_numpy())}, index = df.index)\n",
    "\n",
    "#Returns true if there are no marked genes for >buffer bp upstream of gene, else returns false\n",
    "#Similar to an operon, but naming it 'unit' to emphasize that it is an artificial object\n",
    "def first_in_unit (gene, df, buffer):\n",
    "    strand = int(df[df['Name']==gene_name]['Strand'])\n",
    "    sense_df = df[df['Strand'] == strand].reset_index()\n",
    "    gene_info = sense_df[sense_df['Name']==gene_name]\n",
    "    ind = gene_info.index\n",
    "    if strand == 1:\n",
    "        pass\n",
    "    if strand == -1:\n",
    "        pass\n",
    "    \n",
    "def get_wraparound_string (start, end, genome):\n",
    "    if end > start:\n",
    "        return genome[start:end]\n",
    "    else:\n",
    "        return genome[start:(len(genome)-1)] + genome[0 : end]\n",
    "    \n",
    "\n",
    "def get_upstream_seqs_partial (single_sense_df, buffer, genome):\n",
    "    ss_df = single_sense_df.sort_values(by  = \"StartV3\")\n",
    "    ss_df = ss_df.reset_index()\n",
    "    \n",
    "    upstream = []\n",
    "    for ind in ss_df.index:        \n",
    "        curr_start = ss_df.loc[ind, \"StartV3\"]\n",
    "        prev_end = ss_df.loc[(ind-1) % len(fwd), \"EndV3\"]\n",
    "        if (curr_start - prev_end) % len(genome) < buffer:\n",
    "            upstream.append(None)\n",
    "        else:\n",
    "            upst_start = (curr_start - buffer) % len(genome)\n",
    "            upstream_seq = get_wraparound_string(upst_start, curr_start, genome)\n",
    "            if ss_df.loc[ind, \"Strand\"] == 1:\n",
    "                upstream.append(upstream_seq)\n",
    "            elif ss_df.loc[ind, \"Strand\"] == -1:\n",
    "                upstream.append(reverse_complement (upstream_seq))\n",
    "    return pd.Series(upstream, index = single_sense_df.index)\n",
    "\n",
    "def get_upstream_seqs_full (df, buffer, genome):\n",
    "    upstream = pd.Series([])\n",
    "    for strand in df['Strand'].unique():\n",
    "        upstream = upstream.append(get_upstream_seqs_partial(df[df['Strand']==strand], buffer, genome))\n",
    "    upstream.name = str(buffer) + \"bp upstream sequence\"\n",
    "    return upstream\n",
    "\n",
    "def normalize (exp_data):\n",
    "    exp_data = np.array(exp_data)\n",
    "    mean = np.mean(exp_data)\n",
    "    stdev = np.std(exp_data)\n",
    "    return list((exp_data - mean) / stdev)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "exp_columns = ['LBexp_1_hyb25350202',\n",
    "'LBexp_2_hyb25368302',\n",
    "'LBexp_3_hyb25348702',\n",
    "'LBtran_1_hyb25350002',\n",
    "'LBtran_2_hyb25350302',\n",
    "'LBtran_3_hyb25350802',\n",
    "'LBstat_1_hyb25350502',\n",
    "'LBstat_2_hyb25350102',\n",
    "'LBstat_3_hyb25350702',\n",
    "'LBGexp_1_hyb25351002',\n",
    "'LBGexp_2_hyb25350402',\n",
    "'LBGexp_3_hyb25350602',\n",
    "'LBGtran_2_hyb27737502',\n",
    "'LBGtran_3_hyb27061402',\n",
    "'LBGstat_1_hyb27738002',\n",
    "'LBGstat_2_hyb27737202',\n",
    "'LBGstat_3_hyb27737302',\n",
    "'S0_1_hyb28386102',\n",
    "'S0_2_hyb28384902_01',\n",
    "'S0_3_hyb29584102',\n",
    "'S1_1_hyb28384002',\n",
    "'S1_2_hyb28358602',\n",
    "'S1_3_hyb29569102',\n",
    "'S2_1_hyb28345202_01',\n",
    "'S2_2_hyb28438702_01',\n",
    "'S2_3_hyb29584002',\n",
    "'S3_1_hyb28353902',\n",
    "'S3_2_hyb28407302_01',\n",
    "'S3_3_hyb29593702',\n",
    "'S4_1_hyb28353602_01',\n",
    "'S4_2_hyb29633702',\n",
    "'S4_3_hyb29593902',\n",
    "'S5_1_hyb28387702',\n",
    "'S5_2_hyb29581902',\n",
    "'S5_3_hyb29636602',\n",
    "'S6_1_hyb28357602',\n",
    "'S6_2_hyb29634602',\n",
    "'S6_3_hyb29637702',\n",
    "'S6/t_1_hyb39203702',\n",
    "'S6/t_2_hyb42359702',\n",
    "'S6/t_2_hyb39204702',\n",
    "'S6_4_hyb43477102',\n",
    "'S6_5_hyb43203402',\n",
    "'S6_6_hyb43259002',\n",
    "'S7_4_hyb43261002',\n",
    "'S7_5_hyb43260902',\n",
    "'S7_6_hyb43477202',\n",
    "'S8_4_hyb43256702',\n",
    "'S8_5_hyb43271102',\n",
    "'G135_1_hyb33853802',\n",
    "'G135_3_hyb33849902',\n",
    "'G150_1_hyb33856402',\n",
    "'G150_2_hyb33873202',\n",
    "'G150_3_hyb29750102',\n",
    "'G180_1_hyb29759602',\n",
    "'G180_2_hyb29634102',\n",
    "'G180_3_hyb33853902',\n",
    "'C30_1_hyb42359802',\n",
    "'C30_2_hyb42068102',\n",
    "'C90_1_hyb39425302',\n",
    "'C90_2_hyb39202602',\n",
    "'BI_1_hyb32835602',\n",
    "'BI_2_hyb32823102',\n",
    "'BC_2_hyb32831902',\n",
    "'BC/t_1_hyb39182802',\n",
    "'BC/t_2_hyb39226502',\n",
    "'BC_1_hyb32815402',\n",
    "'Sw_1_hyb32815602',\n",
    "'Sw_2_hyb32815502',\n",
    "'BT_1_hyb32815202',\n",
    "'BT_2_hyb32815702',\n",
    "'B36_1_hyb32830302',\n",
    "'B36_2_hyb32815102',\n",
    "'B60_1_hyb32815302',\n",
    "'B60_2_hyb32835802',\n",
    "'M9exp_1_hyb27733402',\n",
    "'M9exp_2_hyb27733502',\n",
    "'M9exp_3_hyb27732902',\n",
    "'M9tran_1_hyb27733002',\n",
    "'M9tran_2_hyb27733202',\n",
    "'M9tran_3_hyb27728702',\n",
    "'M9stat_1_hyb27732502',\n",
    "'M9stat_2_hyb27732802',\n",
    "'M9stat_3_hyb27732702',\n",
    "'M+G_1_hyb15402302',\n",
    "'M+G_2_hyb15346902',\n",
    "'M+G_3_hyb15354202',\n",
    "'Glu_1_hyb15403702',\n",
    "'Glu_2_hyb15349202',\n",
    "'Glu_3_hyb15349002',\n",
    "'Fru_1_hyb15354002',\n",
    "'Fru_2_hyb15345702',\n",
    "'Fru_3_hyb15352002',\n",
    "'Mal_1_hyb15352102',\n",
    "'Mal_2_hyb15350202',\n",
    "'Mal_3_hyb15356102',\n",
    "'G+S_1_hyb15362102',\n",
    "'G+S_2_hyb15386002',\n",
    "'G+S_3_hyb15356002',\n",
    "'Pyr_1_hyb15521602',\n",
    "'Pyr_2_hyb15390702',\n",
    "'Pyr_3_hyb15522102',\n",
    "'Gly_1_hyb15517802',\n",
    "'Gly_2_hyb15356202',\n",
    "'Gly_3_hyb15358002',\n",
    "'Glucon_1_hyb15409902',\n",
    "'Glucon_2_hyb15416802',\n",
    "'Glucon_3_hyb15525402',\n",
    "'aero_1_hyb32848902',\n",
    "'aero_2_hyb32848402',\n",
    "'aero_3_hyb32839402',\n",
    "'nit_1_hyb32848302',\n",
    "'nit_2_hyb32848802',\n",
    "'nit_3_hyb32848702',\n",
    "'ferm_1_hyb32835902',\n",
    "'ferm_2_hyb32848502',\n",
    "'ferm_3_hyb32875102',\n",
    "'T-5.40H_1_hyb32860802',\n",
    "'T-4.40H_1_hyb32876002',\n",
    "'T-3.40H_1_hyb32860502',\n",
    "'T-2.40H_1_hyb32875702',\n",
    "'T-1.40H_1_hyb32874902',\n",
    "'T-1.10H_1_hyb32875902',\n",
    "'T-0.40H_1_hyb32860902',\n",
    "'T0.0H_1_hyb32860102',\n",
    "'T0.30H_1_hyb32876202',\n",
    "'T1.0H_1_hyb34436402',\n",
    "'T1.30H_1_hyb34528602',\n",
    "'T2.0H_1_hyb34508402',\n",
    "'T2.30H_1_hyb34549502',\n",
    "'T3.0H_1_hyb34489302',\n",
    "'T3.30H_1_hyb34489502',\n",
    "'T4.0H_1_hyb34533902',\n",
    "'T5.0H_1_hyb34489402',\n",
    "'HPh_1_hyb11946602',\n",
    "'HPh_2_hyb11953802',\n",
    "'HPh_3_hyb11955202',\n",
    "'LPh_1_hyb11962902',\n",
    "'LPh_2_hyb11953902',\n",
    "'LPh_3_hyb11971902',\n",
    "'LPhT_1_hyb12085002',\n",
    "'LPhT_2_hyb11955302',\n",
    "'LPhT_3_hyb11955002',\n",
    "'SMMPr_1_hyb25347902',\n",
    "'SMMPr_2_hyb25348202',\n",
    "'SMMPr_3_hyb25348302',\n",
    "'HiTm_1_hyb25345002',\n",
    "'HiTm_2_hyb25344702',\n",
    "'HiTm_3_hyb25347002',\n",
    "'LoTm_1_hyb25348402',\n",
    "'LoTm_2_hyb25348102',\n",
    "'LoTm_3_hyb25348002',\n",
    "'HiOs_1_hyb25346802',\n",
    "'HiOs_2_hyb25344902',\n",
    "'HiOs_3_hyb25347502',\n",
    "'BMM_1_hyb25347702',\n",
    "'BMM_2_hyb25347602',\n",
    "'BMM_3_hyb25344402',\n",
    "'Heat_1_hyb25340602',\n",
    "'Heat_2_hyb25344602',\n",
    "'Heat_3_hyb25341002',\n",
    "'Cold_1_hyb25347802',\n",
    "'Cold_2_hyb25344502',\n",
    "'Cold_3_hyb25340802',\n",
    "'Etha_1_hyb25348602',\n",
    "'Etha_2_hyb25340502',\n",
    "'Etha_3_hyb25340702',\n",
    "'SMM_1_hyb25347102',\n",
    "'SMM_2_hyb25346902',\n",
    "'SMM_3_hyb25346702',\n",
    "'Salt_1_hyb25345102',\n",
    "'Salt_2_hyb25344802',\n",
    "'Salt_3_hyb25347202',\n",
    "'Mt0_1_hyb27737602',\n",
    "'Mt0_2_hyb27737802',\n",
    "'Mt0_3_hyb28385002',\n",
    "'M0t45_1_hyb27737402',\n",
    "'M0t45_2_hyb28386202',\n",
    "'M0t45_3_hyb28358702',\n",
    "'M0t90_1_hyb27737702',\n",
    "'M0t90_2_hyb28444802',\n",
    "'M0t90_3_hyb28387802_01',\n",
    "'M40t45_1_hyb25414802',\n",
    "'M40t45_2_hyb28390602',\n",
    "'M40t45_3_hyb28463502',\n",
    "'M40t90_1_hyb27061602',\n",
    "'M40t90_2_hyb28384202_01',\n",
    "'M40t90_3_hyb28384802_01',\n",
    "'dia0_1_hyb25321202',\n",
    "'dia0_2_hyb25340102',\n",
    "'dia0_3_hyb29581702',\n",
    "'dia5_1_hyb25337802',\n",
    "'dia5_2_hyb25340002',\n",
    "'dia5_3_hyb25348502',\n",
    "'dia15_1_hyb25321902',\n",
    "'dia15_2_hyb25339602',\n",
    "'dia15_3_hyb25339402',\n",
    "'Diami_1_hyb27744802',\n",
    "'Diami_2_hyb32848002',\n",
    "'Diami_3_hyb27744702',\n",
    "'Paraq_1_hyb29583002',\n",
    "'Paraq_2_hyb32848102',\n",
    "'Paraq_3_hyb29634702',\n",
    "'H2O2_1_hyb32848202',\n",
    "'H2O2_2_hyb32844902',\n",
    "'H2O2_3_hyb27746602',\n",
    "'Oxctl_1_hyb27746402',\n",
    "'Oxctl_2_hyb27744902',\n",
    "'Oxctl_3_hyb32839302',\n",
    "'GM-0.1_1_hyb10236302',\n",
    "'GM-0.2_1_hyb10240202',\n",
    "'GM+5_1_hyb10240302',\n",
    "'GM+15_1_hyb10242502',\n",
    "'GM+25_1_hyb17179302',\n",
    "'GM+45_1_hyb10244302',\n",
    "'GM+60_1_hyb10244602',\n",
    "'GM+90_1_hyb10319902',\n",
    "'GM+120_1_hyb10322702',\n",
    "'GM+150_1_hyb17158702',\n",
    "'GM-0.1_2_hyb13973302',\n",
    "'GM-0.2_2_hyb17239902',\n",
    "'GM+5_2_hyb10211402',\n",
    "'GM+10_2_hyb10216402',\n",
    "'GM+15_2_hyb10226802',\n",
    "'GM+25_2_hyb16959802',\n",
    "'GM+45_2_hyb10229502',\n",
    "'GM+60_2_hyb10235002',\n",
    "'GM+90_2_hyb20971602',\n",
    "'GM+150_2_hyb10236602',\n",
    "'GM-0.1_3_hyb16991902',\n",
    "'GM-0.2_3_hyb17183602',\n",
    "'GM+5_3_hyb10404402',\n",
    "'GM+10_3_hyb10410902',\n",
    "'GM+15_3_hyb10216702',\n",
    "'GM+25_3_hyb10217102',\n",
    "'GM+45_3_hyb10218302',\n",
    "'GM+60_3_hyb10219302',\n",
    "'GM+90_3_hyb16981502',\n",
    "'GM+120_3_hyb20975802',\n",
    "'GM+150_3_hyb16960402',\n",
    "'MG-0.1_1_hyb10410402',\n",
    "'MG+5_1_hyb10226702',\n",
    "'MG+10_1_hyb16976802',\n",
    "'MG+15_1_hyb17176602',\n",
    "'MG+25_1_hyb17138402',\n",
    "'MG+45_1_hyb10229802',\n",
    "'MG+120_1_hyb10237002',\n",
    "'MG+150_1_hyb17170002',\n",
    "'MG-0.1_2_hyb10238002',\n",
    "'MG-0.2_2_hyb10331702',\n",
    "'MG+t5_2_hyb10374502',\n",
    "'MG+10_2_hyb10217502',\n",
    "'MG+15_2_hyb17000602',\n",
    "'MG+25_2_hyb10218502',\n",
    "'MG+45_2_hyb10218602',\n",
    "'MG+60_2_hyb10218902',\n",
    "'MG+90_2_hyb10219102',\n",
    "'MG+120_2_hyb17176202',\n",
    "'MG-0.1_3_hyb10223302',\n",
    "'MG-0.2_3_hyb10226602',\n",
    "'MG+5_3_hyb20974702',\n",
    "'MG+10_3_hyb10326302',\n",
    "'MG+15_3_hyb24698102',\n",
    "'MG+25_3_hyb14627102',\n",
    "'MG+45_3_hyb14632602',\n",
    "'MG+60_3_hyb17245002',\n",
    "'MG+90_3_hyb14625202',\n",
    "'MG+120_3_hyb14633902_01',\n",
    "'MG+150_3_hyb14630502']\n",
    "unique_exps = np.unique(list(map(lambda x : x.split('_')[0], exp_columns)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "upst20 = get_upstream_seqs_full (genes, 20, genome) \n",
    "upst50 = get_upstream_seqs_full (genes, 50, genome) \n",
    "upst100 = get_upstream_seqs_full (genes, 100, genome) \n",
    "exp_data = merge_exp_columns (genes, unique_exps)\n",
    "norm_exp_data = exp_data['exps'].map(lambda x : normalize(x))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('experiments_order.txt', 'w') as exp_file:\n",
    "    exp_file.writelines(list(map(lambda x : x + '\\n', unique_exps)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_data = pd.DataFrame({'Name': genes['Name'],\n",
    "              '20bp upstream sequence':upst20, \n",
    "              '50bp upstream sequence':upst50, \n",
    "              '100bp upstream sequence' : upst100, \n",
    "              'Expression':exp_data['exps'], \n",
    "              'Norm Expression': norm_exp_data})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_data.to_csv('cleaned_data.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python rdkit",
   "language": "python",
   "name": "my-rdkit-env"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
